OpenRA.Game/Network/UnitOrders.cs

```diff
@@ -81,25 +62,37 @@ internal static void ProcessOrder(OrderManager orderManager, World world, int cl
 					{
 						var client = orderManager.LobbyInfo.ClientWithIndex(clientId);
 
-						if (client != null)
+						if (client == null)
+							break;
+
+						if (world == null)
 						{
-							if (world == null)
-							{
-								if (orderManager.LocalClient != null && client.Team == orderManager.LocalClient.Team)
-									Game.AddChatLine(client.Color, "[Team] " + client.Name, order.TargetString);
-							}
-							else
-							{
-								var player = world.FindPlayerByClient(client);
-								if (player != null && player.WinState == WinState.Lost)
-									Game.AddChatLine(client.Color, client.Name + " (Dead)", order.TargetString);
-								else if ((player != null && world.LocalPlayer != null && player.Stances[world.LocalPlayer] == Stance.Ally) || (world.IsReplay && player != null))
-									Game.AddChatLine(client.Color, "[Team" + (world.IsReplay ? " " + client.Team : "") + "] " + client.Name, order.TargetString);
-								else if ((orderManager.LocalClient != null && orderManager.LocalClient.IsObserver && client.IsObserver) || (world.IsReplay  && client.IsObserver))
-									Game.AddChatLine(client.Color, "[Spectators] " + client.Name, order.TargetString);
-							}
+							if (orderManager.LocalClient != null && client.Team == orderManager.LocalClient.Team)
+								Game.AddChatLine(client.Color, "[Team] " + client.Name, order.TargetString);
+
+							break;
 						}
 
+						// Use only the normal chat when the game is over
+						if (world.IsGameOver)
+						{
+							WriteChatLine(orderManager, world, clientId, order);
+							break;
+						}
+
+						var player = world.FindPlayerByClient(client);
+						var isSpectator = orderManager.LocalClient != null && orderManager.LocalClient.IsObserver;
+						var isLiving = player != null && player.WinState == WinState.Undefined;
+
+						// We count player with WinState.Won here as well
+						var isDead = player != null && player.WinState != WinState.Undefined;
+						var localDead = world.LocalPlayer != null && world.LocalPlayer.WinState != WinState.Undefined;
+
+						if ((isLiving && world.LocalPlayer != null && player.Stances[world.LocalPlayer] == Stance.Ally) || (world.IsReplay && isLiving))
+							Game.AddChatLine(client.Color, "[Team" + (world.IsReplay ? " " + client.Team : "") + "] " + client.Name, order.TargetString);
```

Comment:
It seems like a bad idea to put this directly to the chat instead of going through `WriteChatLine`.  This skips the `ChatMessageMaxLength` checks, for one.

Reply:
`WriteChatLine`?